rskj-core/src/test/java/co/rsk/peg/whitelist/WhitelistStorageProviderImplTest.java

```diff
@@ -0,0 +1,185 @@
+
+        assertEquals(0, actualLockWhitelist.getAll().size());
+    }
+
+    @Test
+    void save_whenLockWhiteListIsNotNull_ShouldReturnSavedValue() {
+        LockWhitelist lockWhitelist = whitelistStorageProvider.getLockWhitelist(activationConfig, networkParameters);
+
+        //make sure the lockWhitelist is empty
+        assertEquals(0, lockWhitelist.getAll().size());
+
+        LockWhitelistEntry oneOffWhiteListEntry =  createOneOffWhiteListEntry();
+        lockWhitelist.put(firstBtcAddress, oneOffWhiteListEntry);
+
+        whitelistStorageProvider.save(activationConfig);
+
+        whitelistStorageProvider = new WhitelistStorageProviderImpl(inMemoryStorage);
+        LockWhitelist actualLockWhitelist = whitelistStorageProvider.getLockWhitelist(activationConfig, networkParameters);
+        assertEquals(1, actualLockWhitelist.getAll().size());
+
+        //Making sure the saved value is correct
+        Map<Address, OneOffWhiteListEntry> whitelistedAddresses = getAddressFromOneOffWhiteListEntryInMemoryStorage();
+        assertTrue(whitelistedAddresses.containsKey(firstBtcAddress));
+    }
+
+    @Test
+    void save_whenIsActiveRSKIP87_ShouldSavedUnlimitedLockWhitelist() {
+        when(activationConfig.isActive(ConsensusRule.RSKIP87)).thenReturn(true);
+        LockWhitelist lockWhitelist = whitelistStorageProvider.getLockWhitelist(activationConfig, networkParameters);
+
+        //make sure the lockWhitelist is empty
+        assertEquals(0, lockWhitelist.getAll().size());
+
+        LockWhitelistEntry oneOffWhiteListEntry =  createOneOffWhiteListEntry();
+        LockWhitelistEntry unlimitedWhiteListEntry = createUnlimitedWhiteListEntry();
+        lockWhitelist.put(firstBtcAddress, oneOffWhiteListEntry);
+        lockWhitelist.put(secondBtcAddress, unlimitedWhiteListEntry);
+
+        whitelistStorageProvider.save(activationConfig);
+
+        whitelistStorageProvider = new WhitelistStorageProviderImpl(inMemoryStorage);
+        LockWhitelist actualLockWhitelist = whitelistStorageProvider.getLockWhitelist(activationConfig, networkParameters);
+        assertEquals(2, actualLockWhitelist.getAll().size());
+
+        //Making sure the saved value is correct
+        Map<Address, OneOffWhiteListEntry> oneOffWhiteListEntryMap = getAddressFromOneOffWhiteListEntryInMemoryStorage();
+        assertTrue(oneOffWhiteListEntryMap.containsKey(firstBtcAddress));
+
+        Map<Address, LockWhitelistEntry> lockWhitelistEntryMap = getAddressFromUnlimitedWhiteListEntryInMemoryStorage();
+        assertTrue(lockWhitelistEntryMap.containsKey(secondBtcAddress));
+    }
+
+    @Test
+    void getLockWhitelist_whenNoEntriesInStorage_shouldReturnZeroEntries() {
+        LockWhitelist actualLockWhitelist = whitelistStorageProvider.getLockWhitelist(activationConfig, networkParameters);
+
+        assertEquals(0, actualLockWhitelist.getAll().size());
+    }
+
+    @Test
+    void getLockWhitelist_whenLockWhitelistIsNotNull_shouldReturnOneEntry() {
```

Comment:
There's a difference between the test name and the actual test scenario.

The scenario tested here is that there is a value in storage and it returns it. lockWhitelist is still null in memory.

To test what happens when lockWhitelist is not null, you should call the storageProvider twice. The second time lockWhitelist will not be null. See how Jere did it https://github.com/rsksmart/rskj/blob/federation-support-refactor-integration/rskj-core/src/test/java/co/rsk/peg/federation/FederationStorageProviderImplTests.java#L662 

Call storageProvider once, get the values. Insert something directly in the storage. Call the storageProvider again, should use the cache so the value inserted directly in storage will not show. Recreate the storageProvider instance, should return both values now

Reply:
Done! Test Case improved. Please let me know if the following accomplishes what we want:

- Changed the name to get closer to what it does.
- This test case is based on these steps:
1. The first time it gets values, `LockWhitelist` is `null` and queries the storage.
2. It should return zero entries as no value is saved in storage. As of here, `LockWhitelist` is `not null`.
3. Save a value directly in storage.
4. The second time it gets values, `LockWhitelist` is `not null` so it doesn't query the storage and get the value from the cache.
5. Return zero entries as it doesn't query in storage.
6. Recreate `whitelistStorageProvider` to ensure it is querying the storage.
7. Return one entry that was saved directly in storage.
8. Making sure the correct value was whitelisted.

_P.S.: I didn't save another entry directly in storage related to `OneOffWhiteListEntry` since it replaces the previous one because they have the same `WhitelistStorageIndexKey`, `LOCK_ONE_OFF`._